precompile/assetsbridge/bridge_out.go

```diff
@@ -0,0 +1,443 @@
+		isBTC = bytes.Equal(
+			common.HexToAddress(evmtypes.BTCTokenPrecompileAddress).Bytes(),
+			inputs.Token.Bytes(),
+		)
+	)
+
+	sdkAmount, err := precompile.TypesConverter.BigInt.ToSDK(inputs.Amount)
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert amount: [%w]", err)
+	}
+	assetsUnlocked, err := m.bridgeKeeper.SaveAssetsUnlocked(
+		context.SdkCtx(),
+		inputs.Recipient,
+		inputs.Token.Bytes(),
+		context.MsgSender().Bytes(),
+		sdkAmount,
+		uint8(inputs.Chain),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to send AssetsUnlocked to bridge: %w", err)
+	}
+
+	err = context.EventEmitter().Emit(
+		NewAssetsUnlockedEvent(
+			assetsUnlocked.UnlockSequence.BigInt(),
+			assetsUnlocked.Recipient,
+			common.HexToAddress(assetsUnlocked.Token),
+			context.MsgSender(),
+			assetsUnlocked.Amount.BigInt(),
+			uint8(assetsUnlocked.Chain), //nolint:gosec // G115: Safe conversion, Chain is validated elsewhere
+		),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to emit AssetsUnlocked event: [%w]", err)
+	}
+
+	switch inputs.Chain {
+	case TargetChainEthereum:
+		if isBTC {
+			err = m.burnBitcoin(context, inputs)
+		} else {
+			err = m.burnERC20(context, inputs)
+		}
+	case TargetChainBitcoin:
+		err = m.burnBitcoin(context, inputs)
+	default:
+		panic(fmt.Sprintf("unreachable, unsupported target chain: %v", inputs.Chain))
+	}
+
+	return precompile.MethodOutputs{err == nil}, err
+}
+
+func (m *BridgeOutMethod) burnERC20(
+	context *precompile.RunContext,
+	inputs *bridgeOutInputs,
+) error {
+	var (
+		sdkCtx   = context.SdkCtx()
+		fromAddr = context.MsgSender()
+	)
+
+	return m.bridgeKeeper.BurnERC20(
+		sdkCtx, inputs.Token.Bytes(), fromAddr.Bytes(), inputs.Amount)
+}
+
+func (m *BridgeOutMethod) burnBitcoin(
+	context *precompile.RunContext,
+	inputs *bridgeOutInputs,
+) error {
+	bridgeAddr := sdk.AccAddress(common.HexToAddress(
+		evmtypes.AssetsBridgePrecompileAddress,
+	).Bytes())
+	senderAddr := sdk.AccAddress(context.MsgSender().Bytes())
+
+	authorization, expiration := m.authzKeeper.GetAuthorization(
+		context.SdkCtx(), bridgeAddr.Bytes(), senderAddr.Bytes(), SendMsgURL,
+	)
+	if authorization == nil {
+		return fmt.Errorf("%s authorization type does not exist or is expired for address %s", SendMsgURL, senderAddr)
+	}
+
+	sendAuth, ok := authorization.(*banktypes.SendAuthorization)
+	if !ok {
+		return fmt.Errorf(
+			"expected authorization to be a %T", banktypes.SendAuthorization{},
+		)
+	}
+
+	sdkAmount, err := precompile.TypesConverter.BigInt.ToSDK(inputs.Amount)
+	if err != nil {
+		return fmt.Errorf("failed to convert amount: [%w]", err)
+	}
+	coin := sdk.Coin{Denom: evmtypes.DefaultEVMDenom, Amount: sdkAmount}
+
+	// now update the authorization to spend for the AssetsBridge
+	msg := banktypes.NewMsgSend(senderAddr.Bytes(), bridgeAddr.Bytes(), sdk.NewCoins(coin))
+	resp, err := sendAuth.Accept(context.SdkCtx(), msg)
+	if err != nil {
+		return fmt.Errorf("couldn't update authorization: %w", err)
```

Comment:
Looks like we have a small mistake here:
```suggestion
		return fmt.Errorf("couldn't check authorization: %w", err)
```